File: app/src/main/java/com/spidroid/starry/viewmodels/StoryViewModel.java
package com.spidroid.starry.viewmodels;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.Query;
import com.spidroid.starry.models.StoryModel;
import com.spidroid.starry.models.UserModel; // قد تحتاجها لجلب معلومات المستخدم
import com.spidroid.starry.repositories.UserRepository; // قد تحتاجها لجلب معلومات المستخدم
import android.util.Log; // لاستخدام Log

import java.util.Date;
import java.util.List;
import java.util.ArrayList;
import java.util.stream.Collectors; // لاستخدام stream().collect()

public class StoryViewModel extends ViewModel {

    private FirebaseAuth auth;
    private FirebaseFirestore db;
    private MutableLiveData<List<StoryModel>> _stories = new MutableLiveData<>();
    public LiveData<List<StoryModel>> getStories() {
        return _stories;
    }

    // LiveData لقصة المستخدم الحالي
    private MutableLiveData<StoryModel> _yourStory = new MutableLiveData<>();
    public LiveData<StoryModel> getYourStory() {
        return _yourStory;
    }

    // LiveData للمستخدم الحالي
    private MutableLiveData<UserModel> _currentUser = new MutableLiveData<>();
    public LiveData<UserModel> getCurrentUser() {
        return _currentUser;
    }

    private UserRepository userRepository; // لإدارة المستخدمين

    public StoryViewModel() {
        auth = FirebaseAuth.getInstance();
        db = FirebaseFirestore.getInstance();
        userRepository = new UserRepository(); // تهيئة UserRepository
    }

    public void fetchStoriesForCurrentUserAndFollowing() {
        String currentUserId = auth.getUid();
        if (currentUserId == null) {
            _stories.setValue(new ArrayList<>());
            return;
        }

        // ابدأ بجلب المستخدمين الذين يتابعهم المستخدم الحالي
        db.collection("users").document(currentUserId)
                .collection("following")
                .get()
                .addOnSuccessListener(followingSnapshot -> {
                    List<String> followingIds = new ArrayList<>();
                    for (com.google.firebase.firestore.DocumentSnapshot doc : followingSnapshot.getDocuments()) {
                        followingIds.add(doc.getId());
                    }
                    followingIds.add(currentUserId); // أضف المستخدم الحالي أيضًا لجلب قصصه

                    if (followingIds.isEmpty()) {
                        _stories.setValue(new ArrayList<>());
                        return;
                    }

                    // الآن، جلب القصص من هؤلاء المستخدمين
                    db.collection("stories")
                            .whereIn("authorId", followingIds) // جلب القصص من المستخدمين الذين يتابعهم
                            .orderBy("createdAt", Query.Direction.DESCENDING)
                            .get()
                            .addOnSuccessListener(storySnapshot -> {
                                List<StoryModel> fetchedStories = new ArrayList<>();
                                for (com.google.firebase.firestore.DocumentSnapshot doc : storySnapshot.getDocuments()) {
                                    StoryModel story = doc.toObject(StoryModel.class);
                                    if (story != null && (story.getExpiresAt() == null || story.getExpiresAt().after(new Date()))) {
                                        fetchedStories.add(story);
                                    }
                                }
                                _stories.setValue(fetchedStories);
                            })
                            .addOnFailureListener(e -> {
                                Log.e("StoryViewModel", "Error fetching stories: " + e.getMessage());
                                _stories.setValue(new ArrayList<>());
                            });
                })
                .addOnFailureListener(e -> {
                    Log.e("StoryViewModel", "Error fetching following list: " + e.getMessage());
                    _stories.setValue(new ArrayList<>());
                });
    }

    public void uploadStory(StoryModel story) {
        // منطق تحميل القصة إلى Firestore
        // ...
    }

    public void markStoryAsViewed(String storyId, String userId) {
        db.collection("stories").document(storyId)
                .update("viewers", com.google.firebase.firestore.FieldValue.arrayUnion(userId))
                .addOnSuccessListener(aVoid -> Log.d("StoryViewModel", "Story viewed by " + userId))
                .addOnFailureListener(e -> Log.e("StoryViewModel", "Error marking story as viewed: " + e.getMessage()));
    }

    // دالة جديدة لجلب قصة المستخدم الحالي
    public void fetchYourStory() {
        String currentUserId = auth.getUid();
        if (currentUserId == null) {
            _yourStory.setValue(null);
            return;
        }

        db.collection("stories")
                .whereEqualTo("authorId", currentUserId)
                .orderBy("createdAt", Query.Direction.DESCENDING)
                .limit(1) // جلب أحدث قصة للمستخدم
                .get()
                .addOnSuccessListener(queryDocumentSnapshots -> {
                    if (!queryDocumentSnapshots.isEmpty()) {
                        StoryModel story = queryDocumentSnapshots.getDocuments().get(0).toObject(StoryModel.class);
                        // تحقق من انتهاء صلاحية القصة
                        if (story != null && (story.getExpiresAt() == null || story.getExpiresAt().after(new Date()))) {
                            _yourStory.setValue(story);
                        } else {
                            _yourStory.setValue(null); // القصة منتهية الصلاحية
                        }
                    } else {
                        _yourStory.setValue(null); // لا توجد قصة للمستخدم
                    }
                })
                .addOnFailureListener(e -> {
                    Log.e("StoryViewModel", "Error fetching your story: " + e.getMessage());
                    _yourStory.setValue(null);
                });
    }

    // دالة لجلب بيانات المستخدم الحالي (مثل صورة الأفاتار)
    public void fetchCurrentUser() {
        String currentUserId = auth.getUid();
        if (currentUserId != null) {
            userRepository.getUserById(currentUserId).observeForever(user -> {
                _currentUser.setValue(user);
            });
        } else {
            _currentUser.setValue(null);
        }
    }
}

File: app/src/main/java/com/spidroid/starry/viewmodels/StoryViewModel.kt
package com.spidroid.starry.viewmodels

import android.util.Log
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.ktx.toObject
import com.spidroid.starry.models.StoryModel
import com.spidroid.starry.models.UserModel
import com.spidroid.starry.repositories.UserRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.tasks.await
import java.util.Date
import javax.inject.Inject

// Sealed class to represent the different states of the Story Feed UI
sealed class StoryFeedState {
    object Loading : StoryFeedState()
    data class Success(
        val storyPreviews: List<StoryPreview>,
        val hasMyActiveStory: Boolean
    ) : StoryFeedState()
    data class Error(val message: String) : StoryFeedState()
}

// Data class to hold combined data for the adapter
data class StoryPreview(
    val user: UserModel,
    val hasUnseenStories: Boolean
)

@HiltViewModel
class StoryViewModel @Inject constructor(
    private val auth: FirebaseAuth,
    private val db: FirebaseFirestore,
    private val userRepository: UserRepository
) : ViewModel() {

    private val _storyFeedState = MutableStateFlow<StoryFeedState>(StoryFeedState.Loading)
    val storyFeedState: StateFlow<StoryFeedState> = _storyFeedState

    private val _currentUser = MutableLiveData<UserModel?>()
    val currentUser: LiveData<UserModel?> get() = _currentUser

    companion object {
        private const val TAG = "StoryViewModel"
    }

    init {
        fetchCurrentUser()
        fetchStoriesForCurrentUserAndFollowing()
    }

    // --- تم تعديل هذه الدالة بالكامل ---
    fun fetchCurrentUser() {
        auth.currentUser?.uid?.let { userId ->
            viewModelScope.launch {
                try {
                    // استدعاء الدالة الصحيحة من الـ Repository
                    val user = userRepository.getUser(userId)
                    _currentUser.postValue(user)
                } catch (e: Exception) {
                    Log.e(TAG, "Failed to fetch current user", e)
                    _currentUser.postValue(null)
                }
            }
        }
    }

    fun fetchStoriesForCurrentUserAndFollowing() {
        val currentUserId = auth.currentUser?.uid ?: run {
            _storyFeedState.value = StoryFeedState.Error("User not authenticated")
            return
        }

        viewModelScope.launch {
            _storyFeedState.value = StoryFeedState.Loading
            try {
                val followingSnapshot = db.collection("users").document(currentUserId).get().await()
                val userModel = followingSnapshot.toObject<UserModel>()
                val followingIds = userModel?.following?.keys?.toMutableList() ?: mutableListOf()

                val storyPreviews = mutableListOf<StoryPreview>()
                if (followingIds.isNotEmpty()) {
                    val chunks = followingIds.chunked(10)

                    for (chunk in chunks) {
                        val storiesSnapshot = db.collection("stories")
                            .whereIn("userId", chunk)
                            .whereGreaterThan("expiresAt", Date())
                            .get()
                            .await()

                        val storiesByUserId = storiesSnapshot.documents.mapNotNull { it.toObject<StoryModel>() }
                            .groupBy { it.userId }

                        for (userIdWithStory in storiesByUserId.keys) {
                            val userDoc = db.collection("users").document(userIdWithStory!!).get().await()
                            val user = userDoc.toObject<UserModel>()
                            user?.let {
                                val viewedStoriesSnapshot = db.collection("users").document(currentUserId)
                                    .collection("viewed_stories").get().await()
                                val viewedStoryIds = viewedStoriesSnapshot.documents.map { it.id }.toSet()

                                val hasUnseen = storiesByUserId[userIdWithStory]?.any { !viewedStoryIds.contains(it.storyId) } == true
                                storyPreviews.add(StoryPreview(user = it, hasUnseenStories = hasUnseen))
                            }
                        }
                    }
                }

                val myStoriesSnapshot = db.collection("stories")
                    .whereEqualTo("userId", currentUserId)
                    .whereGreaterThan("expiresAt", Date())
                    .limit(1)
                    .get()
                    .await()
                val hasMyStory = !myStoriesSnapshot.isEmpty

                _storyFeedState.value = StoryFeedState.Success(storyPreviews, hasMyStory)

            } catch (e: Exception) {
                Log.e(TAG, "Error fetching stories feed", e)
                _storyFeedState.value = StoryFeedState.Error("Failed to load stories: ${e.message}")
            }
        }
    }
}